src/cmd/services/m3dbnode/config/bootstrap.go

```diff
@@ -43,16 +45,89 @@ import (
 var (
 	// defaultNumProcessorsPerCPU is the default number of processors per CPU.
 	defaultNumProcessorsPerCPU = 0.125
+
+	// default order in which bootstrappers are run
+	// (run in ascending order of precedence).
+	defaultOrderedBootstrappers = []string{
+		// Filesystem bootstrapping must be first.
+		bfs.FileSystemBootstrapperName,
+		// Peers and commitlog must come before the uninitialized topology bootrapping.
+		commitlog.CommitLogBootstrapperName,
+		peers.PeersBootstrapperName,
+		uninitialized.UninitializedTopologyBootstrapperName,
+	}
+
+	// bootstrapper order where peers is prefered over commitlog.
+	preferPeersOrderedBootstrappers = []string{
+		// Filesystem bootstrapping must be first.
+		bfs.FileSystemBootstrapperName,
+		// Prefer peers over commitlog.
+		peers.PeersBootstrapperName,
+		commitlog.CommitLogBootstrapperName,
+		uninitialized.UninitializedTopologyBootstrapperName,
+	}
+
+	validBootstrapModes = []BootstrapMode{
+		DefaultBootstrapMode,
+		PreferPeersBootstrapMode,
+	}
+
+	errReadBootstrapModeInvalid = errors.New("bootstrap mode invalid")
+)
+
+// BootstrapMode defines the mode in which bootstrappers are run.
+type BootstrapMode uint
+
+const (
+	// DefaultBootstrapMode executes bootstrappers in default order.
+	DefaultBootstrapMode BootstrapMode = iota
+	// PreferPeersBootstrapMode executes peers before commitlog bootstrapper.
+	PreferPeersBootstrapMode
 )
 
+// UnmarshalYAML unmarshals an BootstrapMode into a valid type from string.
+func (m *BootstrapMode) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var str string
+	if err := unmarshal(&str); err != nil {
+		return err
+	}
+
+	// If unspecified, use default mode.
+	if str == "" {
+		*m = DefaultBootstrapMode
+		return nil
+	}
+
+	strs := make([]string, 0, len(validBootstrapModes))
+	for _, valid := range validBootstrapModes {
+		if str == valid.String() {
+			*m = valid
+			return nil
+		}
+		strs = append(strs, "'"+valid.String()+"'")
+	}
+	return fmt.Errorf("invalid BootstrapMode '%s' valid types are: %s",
```

Comment:
nit: don't think you need to alloc the strs slice since each `BootstrapMode` is already an `io.Stringer`, so it should look the same with Errorf formatting

Reply:
Unfortunately the `strings.Join` takes in `[]string` and not `[]io.Stringer` so think we still need this unless we omit the valid ones from the error.

Reply:
Changed to just add directly the array as part of the log instead of joining.